src/System.Runtime/src/System/HashCode.cs

```diff
@@ -0,0 +1,250 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Runtime;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    public unsafe struct HashCode
+    {
+        private const uint Queue = 4;
+        private const uint Length = 7;
+
+        private const uint Prime1 = 2654435761U;
+        private const uint Prime2 = 2246822519U;
+        private const uint Prime3 = 3266489917U;
+        private const uint Prime4 = 0668265263U;
+        private const uint Prime5 = 0374761393U;
+        
+        // 8x8 = 64 bytes = 1 x86 cache line
+        // +----+----+----+----+----+----+----+----+
+        // | 00 | 01 | 02 | 03 | 04 | 05 | 06 | 07 |
+        // | xxHash State      | Queue        | Le |
+        // +----+----+----+----+----+----+----+----+
+        private fixed uint _state[8];
+        
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        [TargetedPatchingOptOut("Performance critical for inlining across NGen images.")]
+        private void Add(int value)
+        {
+            unchecked
+            {
+                var val = (uint)value;
+
+                fixed (uint* state = _state)
+                {
+                    var queue = state + Queue;
+                    var position = state[Length] & 0x3;
+
+                    if (position < 3)
+                    {
+                        queue[position] = val;
+                    }
+                    else
+                    {
+
+                        var sentinel =
+                            state[0] | state[1] | state[2] | state[3] | // any of accumulators set?
+                            (state[Length] ^ position); // length greater than 3?
+
+                        if (sentinel == 0)
+                        {
+                            // Initialize
+
+                            state[0] = Prime1 + Prime2;
```

Comment:
To avoid HashCode consumers to confuse the hash-code to be suitable for persistence, the initialization should make use of a random, per-app-domain, seed. For XX32, The seed is added to the first three state-vars (`state[0] += seed`, same for idx 1 and 2), and being subtracted from for the last (`state[3] = seed - state[3]`).

Reply:
No problem. I still have the seed code in my polyfill.

This would break the test vector test (which the build server is having problems with anyway). I can't think of any other way to test the output of this function. Suggestions?

Reply:
At my [prototype project Haschisch.Kastriert](https://github.com/gimpf/Haschisch.Kastriert) I implemented a version where the seed could be specified explicitly. The tests would use a known seed, and the overloads for the default seed would use random numbers (from the security.cryptography randomnumbergenerator).